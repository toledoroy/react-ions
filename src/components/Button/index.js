export ButtonAnchor from './ButtonAnchor'
export ButtonConfirmation from './ButtonConfirmation'
export default from './Button'
