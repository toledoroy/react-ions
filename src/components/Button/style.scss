@import '../../styles/common/mixins';
@import '../../styles/common/colors';
@import '../../styles/common/screen';

.btn {
  @include vertical-align('inline');
  @include button-context($primary-1, $white);
  border: 1px solid rgba(37, 50, 64, 0.2);
  border-radius: 3px;
  cursor: pointer;
  font-size: 14px;
  font-weight: 200;
  line-height: 1.1;
  margin-right: 12px;
  outline: none;
  padding: 8px 16px;
  text-decoration: none;
  transform: translateY(0);
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;

  &:hover,
  &:focus {
    box-shadow: 0 2px 8px 0 rgba($primary-3, 0.2);
  }

  &:active {
    box-shadow: inset 0 2px 4px 0 rgba($primary-3, 0.1), 0 2px 8px 0 rgba($primary-3, 0.1);
  }

  svg {
    fill: $white;
  }
  
  em {
    align-items: center;
    display: flex;
    font-style: normal;
    justify-content: center;
  }

  &.loading > em,
  &.loading + svg {
    opacity: 0;
  }

  &.fill {
    box-sizing: border-box;
    justify-content: center;
    width: 100%;

    em {
      // This is not needed for any browser except for Safari 10 and below (on fill items)
      justify-content: center;
      width: 100%;
    }
  }

  // Button size(s)
  &.lg  {
    padding: 11px 8px;
    position: relative;
  }

  // Button types
  &.success {
    @include button-context($green, $white);
  }

  &.danger {
    @include button-context($danger, $white);
  }

  &.warning {
    @include button-context($warning, $white);
  }

  &.danger-alt {
    @include button-context(transparent, $danger);
    border: 1px solid rgba($danger, 0.6);
  }

  &.flat {
    @include button-context($neutral-2, $neutral-4);
  }

  &.info {
    @include button-context($neutral-4, $white);
  }

  &.secondary {
    @include button-context($white, $primary-4);
    border: 1px solid $neutral-3;
  }

  &.plain {
    @include button-context(transparent, $primary-4);
    border: 1px solid transparent;
    font-weight: normal;

    &:hover,
    &:focus {
      box-shadow: none;
    }
  }

  &.plain-light {
    @include button-context(transparent, lighten($primary-4, 40%));
    border: 1px solid transparent;

    &:hover,
    &:focus {
      box-shadow: none;
    }
  }

  &.inverted {
    @include button-context(transparent, $primary-1);
    border: 1px solid rgba($primary-1, 0.6);
  }

  // To handle inline icons
  svg + span,
  span + svg {
    margin-left: 6px;
  }

  @media #{$media-small-minus} {
    &.collapse {
      span, em > span {
        display: none;
      }
      svg {
        margin: 0;
      }
    }
  }
}

.confirmation-wrapper {
  align-items: center;
  display: inline-flex;
  flex-direction: column;
  position: relative;
  justify-content: center;
}

.confirmation-overlay {
  background-color: $white;
  box-sizing: border-box;
  box-shadow: 0 1px 2px 1px rgba(0, 0, 0, 0.2);
  border-radius: 3px;
  position: absolute;
  visibility: hidden;
  opacity: 0;
  top: calc(100% + 10px);
  padding: 8px;
  transition: opacity .25s ease;
  width: 160px;
  z-index: 1;

  &.wide {
    width: 200px;
  }

  &.visible {
    visibility: visible;
    opacity: 1;
  }

  .confirmation-text {
    font-size: 14px;
    font-weight: 200;
    color: #233040;
    margin: 8px 8px 24px 8px;
    display: block;
  }

  .confirmation-approve-btn {
    padding: 0 16px;
  }

  .button-wrapper {
    display: flex;
    justify-content: flex-end;

    > button {
      position: relative;
      margin-right: 8px;

      &:last-child {
        margin-right: 0;
      }
    }
  }

  > em {
    border-style: solid;
    border-width: 0 5px 5px 5px;
    border-color: transparent transparent $white transparent;
    content: '';
    display: inline-block;
    filter: drop-shadow(0px -1px 1px rgba(0, 0, 0, .2));
    position: absolute;
    top: -5px;

    &:after {
      background-color: $white;
      content: '';
      display: block;
      height: .5px;
      position: absolute;
      left: -5.4px;
      top: 5px;
      width: 10px;
    }
  }
}
