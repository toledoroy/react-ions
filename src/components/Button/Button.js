import React from 'react'
import PropTypes from 'prop-types'
import Spinner from 'react-ions/lib/components/Spinner'
import style from './style.scss'
import optclass, { mapOptClass } from '../internal/OptClass'
import colors from '../internal/colors'

const Button = props => {
  const collapseClass = props.collapse ? 'collapse' : null
  const loaderClasses = props.loading ? 'loading' : null
  const btnClasses = optclass(style, [style.btn, props.size, loaderClasses, collapseClass], props.optClass, props.className)

  const styles = mapOptClass(props.optClass, {
    secondary: {
      color: colors.primary4
    },
    warning: {
      color: colors.white
    },
    inverted: {
      color: colors.primary1
    },
    danger: {
      color: colors.white
    },
    'danger-alt': {
      color: colors.danger
    },
    success: {
      color: colors.white
    },
    flat: {
      color: colors.neutral4
    },
    info: {
      color: colors.white
    },
    plain: {
      color: colors.primary4
    },
    'plain-light': {
      color: '#7b96b5'
    },
    default: {
      color: colors.white
    }
  })

  return (
    <button
      type={props.type}
      style={props.style}
      className={btnClasses}
      disabled={props.disabled || props.loading}
      aria-disabled={props.disabled || props.loading}
      onClick={props.onClick}
      onMouseEnter={props.onMouseEnter}
      onMouseLeave={props.onMouseLeave}
      onMouseDown={props.onMouseDown}
      onMouseOut={props.onMouseOut}
      onMouseOver={props.onMouseOver}
      onMouseUp={props.onMouseUp}>
      { props.loading 
        && <Spinner
        loading={props.loading}
        type='spinner-circular'
        backgroundColor={'transparent'}
        color={styles.color || props.loaderColor} 
        size='20'
        />
      }
      <em>{props.children}</em>
    </button>
  )
}

Button.defaultProps = {
  type: 'button'
}

Button.propTypes = {
  /**
   * The size of button.
   */
  size: PropTypes.string,
  /**
   * Whether the button is disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Whether the loading spinner is displayed.
   */
  loading: PropTypes.bool,
  /**
   * Whether to display only an icon on small screens
   */
  collapse: PropTypes.bool,
  /**
   * Optional CSS class(es) to be used for local styles (string or array of strings)
   */
  optClass: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.string
  ]),
  /**
   * A function to be called onClick
   */
  onClick: PropTypes.func,
  /**
   * A function to be called onMouseEnter
   */
  onMouseEnter: PropTypes.func,
  /**
   * A function to be called onMouseLeave
   */
  onMouseLeave: PropTypes.func,
  /**
   * A function to be called onMouseDown
   */
  onMouseDown: PropTypes.func,
  /**
   * A function to be called onMouseOut
   */
  onMouseOut: PropTypes.func,
  /**
   * A function to be called onMouseOver
   */
  onMouseOver: PropTypes.func,
  /**
   * A function to be called onMouseUp
   */
  onMouseUp: PropTypes.func,
  /**
   * The type of button.
   */
  type: PropTypes.string,
  /**
   * A string to allow for inline styles
   */
  style: PropTypes.string,
  /**
   * A class name to be used for local styles or integrations (required to support styled-components)
   **/
  className: PropTypes.string,
  /**
   * A valid css color to set the color of the loader (if applicable).
   **/
  loaderColor: PropTypes.string
}

export default Button
