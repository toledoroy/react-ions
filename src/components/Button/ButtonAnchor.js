import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router'
import style from './style.scss'
import classNames from 'classnames/bind'

const ButtonAnchor = props => {
  const cx = classNames.bind(style)
  const collapseClass = props.collapse ? 'collapse' : null
  const btnAnchorClasses = cx(style.btn, props.optClass, props.size, collapseClass)

  let buttonAnchor

  const handleClick = e => {
    if (props.disabled) {
      e.preventDefault()
    }
  }

  if (props.internal) {
    buttonAnchor = <Link to={props.path} className={btnAnchorClasses} size={props.size} disabled={props.disabled} onClick={handleClick} style={props.style} className={props.className}>{props.children}</Link>
  } else {
    buttonAnchor = <a href={props.path} className={btnAnchorClasses} target={props.target} disabled={props.disabled} onClick={handleClick}>{props.children}</a>
  }
  return buttonAnchor
}

ButtonAnchor.propTypes = {
  /**
   * A string to allow for inline styles
   */
  style: PropTypes.string,
  /**
   * A class name to be used for local styles or integrations (required to support styled-components)
   **/
  className: PropTypes.string,
  /**
   * Optional styles to add to the button.
   */
  optClass: PropTypes.string,
  /**
   * The size of button.
   */
  size: PropTypes.string,
  /**
   * Whether the button is disabled.
   */
  disabled: PropTypes.bool,
  /**
   * A path to pass to the anchor tag.
   */
  path: PropTypes.string,
  /**
   * Whether the link it to an internal page, or external (default)
   */
  internal: PropTypes.bool,
  /**
   * Whether to display only an icon on small screens
   */
  collapse: PropTypes.bool
}

export default ButtonAnchor
