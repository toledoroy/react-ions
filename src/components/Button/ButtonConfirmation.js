import React, { Component} from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames/bind'
import optclass from '../internal/OptClass'
import Button from './Button'
import style from './style.scss'

export class ButtonConfirmation extends Component {
  constructor(props) {
    super(props)
    this.mql = window.matchMedia('(max-width: 992px)')
  }

  static defaultProps = {
    prompt: 'Are you sure?'
  }

  static propTypes = {
    /**
     * Optional styles to add to the button.
     */
    optClass: PropTypes.oneOfType([
      PropTypes.array,
      PropTypes.string
    ]),
    /**
     * The size of button.
     */
    size: PropTypes.string,
    /**
     * Whether the button is disabled.
     */
    disabled: PropTypes.bool,
    /**
     * Whether to display only an icon on small screens
     */
    collapse: PropTypes.bool,
    /**
     * Position of the button on the page for overlay caret orientation when the screen size is tablet-sized or smaller
     */
    position: PropTypes.oneOf(['left', 'right']),
    /**
     * Text that will appear in the overlay prompt.
     */
    prompt: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    /**
     * Function used to pass up the confirmation to the parent component
     */
    handleConfirmation: PropTypes.func
  }

  state = {
    confirmationOverlayOpen: false,
    confirmationOverlayOffset: 0,
    hasBeenOpened: false,
    wide: false
  }

  handleOpen = () => {
    if (!this.props.hasBeenOpened) {
      this.setState({
        hasBeenOpened: true
      }, () => {
        this.handleSetup()
      })
    }
    this.setState({
      confirmationOverlayOpen: !this.state.confirmationOverlayOpen
    })
  }

  handleConfirmation = confirm => {
    this.setState({ confirmationOverlayOpen: false }, () => {
      if (typeof this.props.handleConfirmation === 'function' && confirm) {
        this.props.handleConfirmation(confirm)
      }
    })
  }

  handleSetup = () => {
    const triggerRect = this._trigger.children[0].getBoundingClientRect()
    const overlayRect = this._trigger.querySelector("[class*='confirmation-overlay']").getBoundingClientRect()

    this.setState({
      triggerWidth: triggerRect.width,
      overlayWidth: overlayRect.width
    })
  }

  handleWide = () => {
    if (this.props.prompt.length > 25) {
      this.setState({ wide: true })
    }
  }

  getStyles = () => {
    if (this.props.position === 'right') {
      return {
        right: `${this.state.confirmationOverlayOffset}px`
      }
    }
    if (this.props.position === 'left') {
      return {
        left: `${this.state.confirmationOverlayOffset}px`
      }
    }

    // https://developers.google.com/web/updates/2016/06/absolute-positioned-children
    // Once ^ is supported in Safari and Firefox we can remove this and allow flex box to do its thing
    return {
      left: `-${((this.state.overlayWidth - this.state.triggerWidth) / 2)}px`
    }
  }

  getCaretStyles = () => {
    if (this.props.position === 'right') {
      return { right: `calc(${(this.state.triggerWidth / 2) + 7.5}px)`}
    }
    if (this.props.position === 'left') {
      return { left: `calc(${(this.state.triggerWidth / 2) - 5}px)` }
    }
    return this.state.wide ? { left: '95px' } : { left: '75px' }
  }

  /**
   * To trigger JavaScript changes on a breakpoint
   * @param  {Object} mediaQueryList
   * @return {Boolean} whether the viewport is < or > than max-width 768px
   */
  handleMediaChange = () => {
    this.handleSetup()
  }

  componentDidMount = () => {
    this.mql.addListener(this.handleMediaChange)
    this.handleSetup()
    this.handleWide()
  }

  componentWillUnmount = () => {
    this.mql.removeListener(this.handleMediaChange)
  }

  shouldComponentUpdate = (nextProps, nextState) => {
    if (nextProps.position !== this.props.position) return true
    if (nextState.wide !== this.state.wide) return true
    if (nextState.triggerWidth !== this.state.triggerWidth) return true
    if (nextState.overlayWidth !== this.state.overlayWidth) return true
    if (nextState.confirmationOverlayOpen !== this.state.confirmationOverlayOpen) return true
    if (nextProps.prompt !== this.props.prompt) return true
    if (nextProps.loading !== this.props.loading) return true

    return false
  }

  render = () => {
    const cx = classNames.bind(style)
    const { ...other } = this.props

    const buttonClass = optclass(style, ['confirmation-approve-btn'])
    const overlayPositionClass = this.props.position ? style[this.props.position] : null
    const overlayVisibleClass = this.state.confirmationOverlayOpen ? style['visible'] : null
    const overlayWideClass = this.state.wide ? style['wide'] : null
    const confirmationOverlayClasses = cx(overlayVisibleClass, overlayPositionClass, overlayWideClass, style['confirmation-overlay'])

    return (
      <div ref={trigger => this._trigger = trigger} className={style['confirmation-wrapper']}>
        <Button {...other} disabled={this.state.confirmationOverlayOpen} onClick={this.handleOpen}>
          { this.props.children }
        </Button>
        <div className={confirmationOverlayClasses} style={this.getStyles()}>
          <em style={this.getCaretStyles()}></em>
          <span className={style['confirmation-text']}>{this.props.prompt}</span>
          <div className={style['button-wrapper']}>
            <Button onClick={this.handleConfirmation.bind(this, false)} optClass='danger-alt'>Cancel</Button>
            <Button onClick={this.handleConfirmation.bind(this, true)} optClass={buttonClass}>Yes</Button>
          </div>
        </div>
      </div>
    )
  }
}

export default ButtonConfirmation
