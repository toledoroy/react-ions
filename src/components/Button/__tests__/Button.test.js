import React from 'react'
import Button from '../Button'

describe('Button', () => {
  let wrapper

  it('displays a button', () => {
    expect(shallow(<Button disabled optClass='danger' size='lg'>Test</Button>)).toMatchSnapshot()
  })

  it('displays a disabled button with a loader', () => {
    expect(shallow(<Button loading={true}>Test</Button>)).toMatchSnapshot()
  })

  it('hides text when collapse prop is passed', () => {
    expect(shallow(<Button collapse={true}>Test</Button>)).toMatchSnapshot()
  })

  it('should have an extra class', () => {
    expect(shallow(<Button optClass="test-class" />)).toMatchSnapshot()
  })

  it('should have extra classes', () => {
    const optClasses = ['test-class', 'test-class-2', 'test-class-3']
    expect(shallow(<Button size='lg' optClass={optClasses} />)).toMatchSnapshot()
  })

  it('should have mouse event handlers', () => {
    expect(shallow(<Button onClick={sinon.spy()} onMouseEnter={sinon.spy()} onMouseLeave={sinon.spy()} onMouseDown={sinon.spy()} onMouseOut={sinon.spy()} onMouseOver={sinon.spy()} onMouseUp={sinon.spy()}>Test</Button>)).toMatchSnapshot()
  })

  it('should have touch event handlers', () => {
    expect(shallow(<Button onTouchCancel={sinon.spy()} onTouchEnd={sinon.spy()} onTouchMove={sinon.spy()} onTouchStart={sinon.spy()}>Test</Button>)).toMatchSnapshot()
  })
})
