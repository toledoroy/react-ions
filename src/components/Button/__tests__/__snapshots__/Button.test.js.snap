// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button displays a button 1`] = `
<button
  aria-disabled={true}
  className="btn lg danger"
  disabled={true}
  type="button"
>
  <em>
    Test
  </em>
</button>
`;

exports[`Button displays a disabled button with a loader 1`] = `
<button
  aria-disabled={true}
  className="btn loading"
  disabled={true}
  type="button"
>
  <Spinner
    backgroundColor="transparent"
    loading={true}
    size="20"
    type="spinner-circular"
  />
  <em>
    Test
  </em>
</button>
`;

exports[`Button hides text when collapse prop is passed 1`] = `
<button
  className="btn collapse"
  type="button"
>
  <em>
    Test
  </em>
</button>
`;

exports[`Button should have an extra class 1`] = `
<button
  className="btn test-class"
  type="button"
>
  <em />
</button>
`;

exports[`Button should have extra classes 1`] = `
<button
  className="btn lg test-class test-class-2 test-class-3"
  type="button"
>
  <em />
</button>
`;

exports[`Button should have mouse event handlers 1`] = `
<button
  className="btn"
  onClick={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  type="button"
>
  <em>
    Test
  </em>
</button>
`;

exports[`Button should have touch event handlers 1`] = `
<button
  className="btn"
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  type="button"
>
  <em>
    Test
  </em>
</button>
`;
